### ISSUE TITLE
Switching across Adaptation Sets

### ISSUE THREAD

**Issue by avelad** [google/shaka-player#1071]
According to DASH-IF-IOP v4.1, "3.8 Switching across Adaptation Sets"
http://dashif.org/wp-content/uploads/2017/09/DASH-IF-IOP-v4.1-clean.pdf

"Representations in two or more Adaptation Sets may provide the same content. In addition, the content may be time-aligned and may be offered such that seamless switching across Representations in different Adaptation Sets is possible."
"A content author may signal such seamless switching property across Adaptation Sets by providing a Supplemental Descriptor along with an Adaptation Set with @schemeIdUri set to `urn:mpeg:dash:adaptation-set-switching:2016` and the @value is a comma-separated list of Adaptation Set IDs that may be seamlessly switched to from this Adaptation Set." 

Test stream: https://shaka-player-demo.appspot.com/demo/#asset=https://cdn.demo.anevia.com/live/eds/StarsTV/miradash/StarsTV.mpd;lang=es

The current behaviour is that "enable adaptation" uses differents adaption sets of audio with same language but differents channel configuration.

I expect that ShakaPlayer provides a powerful api for audio languages, the current api only returns the language, I expect that returns audio language, role, channel configuration, ...

**Comment by TheModMaker** [comment]
Unfortunately we removed support for AdaptationSet switching when we added HLS.  We used to honor which AdaptationSets we could switch between, but it didn't work well for HLS.  So we changed to using a heuristic model for determining which streams we can switch between.

Currently, we use the current language and role to determine which streams we can switch between.  For using channel configuration, you already filed #1013 to deal with that.  Once that is done we can talk about how ABR works with that.

**Comment by joeyparrish** [comment]
@avelad, we are working on this issue, but it seems that your test stream does not exhibit the issue.  There are no AdaptationSet IDs and no `urn:mpeg:dash:adaptation-set-switching:2016` descriptors.  We can test and verify with unit tests, but it would be nice to have a test stream from you as well so that we could have more confidence in our fix.

**Comment by avelad** [comment]
@joeyparrish, you can use: https://shaka-player-demo.appspot.com/demo/#asset=https://storage.googleapis.com/wvtemp/kylealexander/packager_test/iop_compliant_mpd_test/llama.mpd;license=https://proxy.uat.widevine.com/proxy?provider=widevine_test;lang=es;build=uncompiled

Manifest: https://storage.googleapis.com/wvtemp/kylealexander/packager_test/iop_compliant_mpd_test/llama.mpd
Widevine server: https://proxy.uat.widevine.com/proxy?provider=widevine_test

**Comment by joeyparrish** [comment]
In that example, there are no observable differences.  What we do today is combine all compatible AdaptationSets.  The `adaptation-set-switching` descriptors in that piece of content say that all AdaptationSets should be combined, so it's not possible to see the bug at work with that content.

How is this bug impacting your application?  Do you have more selective `adaptation-set-switching` descriptors in your case?

**Comment by joeyparrish** [comment]
@avelad, we are having difficulty understanding the negative impact of implied switching between all AdaptationSets.  Can you share with us a use-case or some sample content where this strategy fails or causes problems for your application?

**Comment by avelad** [comment]
@joeyparrish , sorry for late response (I lost the previous notification)

```
<?xml version="1.0" encoding="utf-8"?>

<MPD
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xmlns="urn:mpeg:dash:schema:mpd:2011"
  xsi:schemaLocation="urn:mpeg:dash:schema:mpd:2011 http://standards.iso.org/ittf/PubliclyAvailableStandards/MPEG-DASH_schema_files/DASH-MPD.xsd"
  type="dynamic"
  availabilityStartTime="2015-01-01T00:00:00Z"
  publishTime="2018-02-24T08:54:33.847518Z"
  minimumUpdatePeriod="PT2S"
  minBufferTime="PT0S"
  profiles="urn:mpeg:dash:profile:isoff-live:2011,urn:hbbtv:dash:profile:isoff-live:2012">
  <Period
    id="1"
    start="PT0S">
    <BaseURL>dash/</BaseURL>
    <AdaptationSet
      group="1"
      contentType="audio"
      lang="pl"
      segmentAlignment="true"
      audioSamplingRate="48000"
      mimeType="audio/mp4"
      codecs="mp4a.40.2">
      <AudioChannelConfiguration
        schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011"
        value="2">
      </AudioChannelConfiguration>
      <SegmentTemplate
        timescale="48000"
        initialization="StarsTV-$RepresentationID$-init.mp4"
        media="StarsTV-$RepresentationID$-$Time$.mp4">
        <SegmentTimeline>
        </SegmentTimeline>
      </SegmentTemplate>
      <Representation
        id="audio_20001=96000.track_id=20001"
        bandwidth="96000">
      </Representation>
    </AdaptationSet>
    <AdaptationSet
      group="2"
      contentType="audio"
      lang="pl"
      segmentAlignment="true"
      audioSamplingRate="48000"
      mimeType="audio/mp4"
      codecs="mp4a.40.2">
      <AudioChannelConfiguration
        schemeIdUri="urn:mpeg:dash:23003:3:audio_channel_configuration:2011"
        value="6">
      </AudioChannelConfiguration>
      <SegmentTemplate
        timescale="48000"
        initialization="StarsTV-$RepresentationID$-init.mp4"
        media="StarsTV-$RepresentationID$-$Time$.mp4">
        <SegmentTimeline>
        </SegmentTimeline>
      </SegmentTemplate>
      <Representation
        id="audio_20000=256000.track_id=20000"
        bandwidth="256000">
      </Representation>
    </AdaptationSet>
    <AdaptationSet
      group="3"
      contentType="video"
      par="4:3"
      segmentAlignment="true"
      width="320"
      height="240"
      frameRate="25"
      mimeType="video/mp4"
      codecs="avc1.42C00D"
      startWithSAP="1">
      <SegmentTemplate
        timescale="25"
        initialization="StarsTV-$RepresentationID$-init.mp4"
        media="StarsTV-$RepresentationID$-$Time$.mp4">
        <SegmentTimeline>
        </SegmentTimeline>
      </SegmentTemplate>
      <Representation
        id="video=500000.track_id=10000"
        bandwidth="500000"
        scanType="progressive">
      </Representation>
    </AdaptationSet>
    <AdaptationSet
      group="2"
      contentType="video"
      par="71:40"
      segmentAlignment="true"
      width="852"
      height="480"
      frameRate="25"
      mimeType="video/mp4"
      codecs="avc1.42C01E"
      startWithSAP="1">
      <SegmentTemplate
        timescale="25"
        initialization="StarsTV-$RepresentationID$-init.mp4"
        media="StarsTV-$RepresentationID$-$Time$.mp4">
        <SegmentTimeline>
        </SegmentTimeline>
      </SegmentTemplate>
      <Representation
        id="video=800000.track_id=10001"
        bandwidth="800000"
        scanType="progressive">
      </Representation>
    </AdaptationSet>
  </Period>
</MPD>
```

With the previous MPD (I don't have a shareable url) you can see that there are 2 audio AdaptationSets  with the same languages and properties, but diferentes AudioChannelConfiguration.

According to According to DASH-IF-IOP v4.1, "3.8 Switching across Adaptation Sets", with the previous MPD, the player shouldn't provide adaption between the differentes AdaptationSets  due the missing descriptor `urn:mpeg:dash:adaptation-set-switching:2016`

**Comment by joeyparrish** [comment]
We are discussing the best way to address this.  It seems that if the issue is that we're adapting between audio streams with different numbers of channels, we should consider which is simpler: forbidding adaptation between different numbers of channels, or completely implementing adaptation-set-switching.

**Comment by joeyparrish** [comment]
Would you expect users to be able to manually switch between 2 channels and 6?

**Comment by avelad** [comment]
Yes, I expect that the user can select 2 o 6 channels.

**Comment by joeyparrish** [comment]
Thanks for the the feedback.  We'll work on a solution.

**Comment by joeyparrish** [comment]
I believe that our solution to #1013 will address the immediate need (audio channels), but that this issue is still separate and more general.  I will therefore defer detailed implementation of AdaptationSet switching until v2.5.  Please let me know if you have any concerns with this.

**Comment by avelad** [comment]
No problem, if the https://github.com/google/shaka-player/issues/1013 is in 2.4, I don't have any problem.

**Comment by avelad** [comment]
I'm reviewing and i think that I'll need new methods in the player:
getAudioLanguagesRolesAndChannels()
selectAudioLanguage(language, opt_role, opt_channels)

Is it possible add theses methods?

**Comment by vaage** [comment]
@avelad I see that there has not been any updates on this issue in a while. Is this still an issue for you?

**Comment by avelad** [comment]
Yes, i'm waiting to a solution

**Comment by vaage** [comment]
@avelad I wanted to chime in again to let you know that I am still working on your request. I am sorry it is taking so long. The challenge I am working through is that there is future work that we are trying to balance with this and trying to make sure that our solution for this is strategically aligned with the design changes ahead.

**Comment by avelad** [comment]
Thanks for the information

**Comment by vaage** [comment]
@avelad , once again I am sorry this is taking so long to complete.

I am separating out the "enhancements" you are asking for as their own issue. This will make it easier to track and keep top-of-mind. This will also allow us to close this "bug".

The new issue is #1669. I have done my best to fill out what I think it the bigger issue at hand. Please feel free to add to it if you feel I have missed anything.
